### ISSUE TITLE
Cancellation

### ISSUE THREAD

**Issue by brycelelbach** [executors/futures#28]
We should have an issue for this, although it may not be a Version 1 feature.

When we say cancellation, we do not mean cancellation of work that has already started running. We mean cancellation of stuff that has not yet started running. Thus, it's a request, not a demand.

The best analogy, I think, is `unique_ptr::release`.

**Comment by brycelelbach** [comment]
@griwes please review.

**Comment by griwes** [comment]
I think the better analogy is `uptr.reset();`. Release gives you a resource back, and I don't see a resource you get back when you cancel.

Keep in mind that you need observation of the state, and a way to get temporary ownership when executing the work; this pretty much implies that you have a `shared_ptr` in both the unique and shared future cases. This is quite neat, though, since `shared_ptr` in the future + `weak_ptr` in the object holding a promise (whether that's an explicit object, or a version of packaged task that is itself a promise - doesn't really matter; by the word "promise" I mean the contract held by such an object, not necessarily an actual separate object) give you exactly the semantics that you want - and exactly the safety you need, except for the case of a shared, single-threaded future that we've considered before. (But that case can also be solved if `boost::local_shared_ptr` progresses well into the standard.)

Ideally a destructor would cancel. I can see explicit cancellation being a post-v1 feature, although I'm not sure - because the obvious thing for the destructor is to cancel, and the implementation that uses `shared_ptr`/`weak_ptr` pair gives you that out-of-the-box; it's really been a no-brainer when I implemented that. I'm not sure if that's something we can retrofit after v1, but if we can - then sure. (Not so sure about how to implement not cancelling future in this case, though - by replacing the `weak_ptr` with a `shared_ptr` in promises?)

There's one additional caveat that makes me think we should have this from the start: fire-and-forget tasks. Consider:

```
future<int> f();
f().then([](auto i){ /* do something with side-effects with i */ });
```

If we cancel in destructors, then there's a very high chance the callback doesn't get called. But that's good, that's what the feature is. **However**, if the above piece of code is written against v1, and in v2 we change the behavior of the dtor, the code starts being incorrect.

If we do "destructors cancel", we also need something that I (I think I stole it from Sean Parent) call `detach`, which is actually quite trivial to implement:

```
void detach()
{
    then([keep = _state](auto &&... args){});
}
```

Therefore, in the ideal (for me) world, we'd:
 * cancel in destructors
 * do it from the start so there's no surprises with code written against v1 when compiled with v2
 * have `.detach` (this really is crucial)
 * possibly have `.cancel` (which is really equivalent to a forced move out of the object, albeit with nicer syntax)

(Sorry if this came out as a bit unordered, I've just let the words flow.)

**Comment by LeeHowes** [comment]
> Ideally a destructor would cancel. I can see explicit cancellation being a post-v1 feature, although I'm not sure - because the obvious thing for the destructor is to cancel,

The obvious thing for the destructor is to detach. I mention this to highlight that it isn't obvious at all.

Cancellation is messy because while we don't want to run the continuations, we certainly do want to destroy them and we have to destroy them on the right executor in the general case. We'd have to be very careful as well about the guarantees around cancellation. Do we have to propagate it back up the graph, or only apply it to a given future? If we do apply it to a future then do we have to catch the task before we hand it to an executor or not? What happens if we cancel after it has been handed to an executor? For a then_executor, we hand it over immediately, so at that point only the executor itself can really handle cancellation - and that does not mean it's running, just that the executor has possession of the task. If cancellation blocking and immediate (in which case we have a blocking destructor) or does it enqueue a cancellation somehow? How does an executor associate a queue of cancellations with the currently enqueued work?

I don't think I know how to implement it in the general case.

**Comment by RedBeard0531** [comment]
> have .detach (this really is crucial)

I don't think so . `.detach` is dangerous because it can ignore errors. I actually think all Future types should be marked `[[nodiscard]]` to make it vary clear that they shouldn't just be ignored. This caught *several* potential bugs for us at compile time, even while developing our Future implementation before there were any users. I think the key is that we also need a "terminal" `.then` (ours' is called `.getAsync` because it is semantically closer to a `.get` than a `.then`, even though it takes a callback) that returns void and invokes the callback in a `noexcept` context. I think `.detach` should probably be implemented as something like `.getAsync([&] (auto&& future_or_expected) { assert(!future_or_expected.has_error()); })`, or at least make it UB to `.detach` something that fails to allow that. If we really want a no-op `.detach`, it should probably be a separate `.ignore_value_and_errors()` so the name screams "I'm dangerous!"

Sorry for the screed, but "never silently drop errors" has been a crusade of mine, and something I consider crucial for a well designed Futures API.

**Comment by griwes** [comment]
@LeeHowes I think we are talking about two different things. I'm talking about **continuation** cancellation, i.e.: making sure that the argument to `.then` isn't executed if it is cancelled before (for some values of "before") the internal future mechanism begins the steps to start executing it.

What you seem to be talking about is **scheduling** cancellation, i.e.: make sure that the internal future mechanism doesn't even start.

The second one is much harder, I agree, and I also don't know how to implement it in the general case. However, just continuation cancellation is enough, *I think* - it still saves you work, although it still requires that the executor creates an execution agent; if this is a dealbreaker, then I must not be seeing some specific case where that's how it is. It also maps *very* nicely onto `shared_ptr` semantics (though you need some additional refcounting to detect broken promises).

I am very confident it's implementable, becuase I have done exactly that - with the exception that my executor model doesn't have `then_execute` - but it is quite easy. Last strong ref to the shared state dies => the shared state dies => the continuations die. Or am I misunderstanding `then_execute`? Is the executor itself expected to keep the list of continuations when you call it? If that's the case, then I'll have to rethink some parts of this, but I am certain there's a solution. If the future's shared state is expected to keep all the continuations, though, then the implementation is quite obvious.

So, basically: cancellation feels obvious to me, because it literally maps onto `shared_ptr`/`weak_ptr` semantics.

**Comment by griwes** [comment]
@RedBeard0531 Having dealt with the nonsense of (mostly silently) unhandled errors propagated through promise chains in a certain JavaScript promise library, I'm deeply of the mind that we should always `std::terminate` on unhandled exceptions, just like we do when one escapes a thread. (This probably deserves its own issue? cc @brycelelbach)

This is actually quite easy to do, just track the status in the shared state and terminate in the shared state's destructor if no-one handled the exception.

**Comment by brycelelbach** [comment]
We do have an issue for error handling: #7

**Comment by griwes** [comment]
It seems to be focus on how to attach continuations and error continuations, and not on what to do when an exception is not handled. May still be fine to discuss it there if you wish so, though.

**Comment by LeeHowes** [comment]
I understand that we are talking about continuation cancellation not execution cancellation. I just think that in the extreme, the two are the same. Let's say we greedily batch tasks and have already pushed a queue of tasks over to the GPU - can we pick the last task off the queue that the GPU scheduler is watching to cancel it, even if it hasn't started yet? then_execute may well batch the tasks internally - the whole point would be that we don't have to lazily enqueue as futures trigger, so it can do whatever it needs to and return its own future type. That future could have appropriate cancellation hooks, but I would be concerned about requiring something expensive there.

Further, it might not be cheap for an implementation to work back through a continuation graph, so cancelling anything but the last task might mean significant implementation cost.

In that world, cancellation is really the same as detaching - we make no guarantee anyway, so however many tasks we pull off the end of the queue (with a guarantee that we cancel at least 0) everything else is detached. 

I'm definitely sympathetic to the concern about ignoring failures, so [[nodiscard]] makes a lot of sense. We could also say that some futures are [[nodiscard]]. An executor that terminates on exception, or only allows noexcept continuations, could return a discardable future.

What we do with unhandled exceptions was something Michael Wong and others were looking at as I understand it. My goal is to keep the number of unhandled exceptions as close to 0 as possible through concrete mechanisms for passing them. A detached future is clearly an unhandled exception with no future to package it in, and no callback to process it, so I think that comes under the purview of that effort. In that world, I think we have to consider if we want to support an explicit detach on a [[nodiscard]] future, and if we do does it detach and allow exceptions to fall into a hole, or detach and treat exceptions as leaking.

**Comment by griwes** [comment]
> In that world, cancellation is really the same as detaching - we make no guarantee anyway, so however many tasks we pull off the end of the queue (with a guarantee that we cancel at least 0) everything else is detached.

This is true, and it has actually convinced me that "Not Version 1" is enough for cancellation. I'll weep a bit about the default action in destructor, though you've made me realize it won't be too hard of a weep.

`[[nodiscard]]` is an attribute and as such, we can't really consider it to be a proper way to deal with error handling. I wholeheartedly think that just terminating when you encounter such a failure situation is fine. I expect `.detach()` to be such a rare (but still extremely important for me!) case, that we can require the user to actually be careful about what we are doing. We could *try* something like "you are only allowed to do this on a future that follows a `noexcept` continuation" or something similar, but I am very skeptical whether this makes sense.

**Comment by LeeHowes** [comment]
I talked to others at FB about this and we are in favour of explicit detach. It would be best if we can enforce it at compile time in some way.

The question then is what should the destructor do if still attached and incomplete? Terminate? Throw? Maybe throw is fine in this rare case because the option is terminate, so a double-throw that terminates is no worse, and in the best case marginally more controlled?

**Comment by LeeHowes** [comment]
Terminate (+nodiscard in this case as it is an obvious return value) would be consistent with std::thread.

**Comment by griwes** [comment]
If we do explicit detach (and explicit cancelation, eventually), then I'm in favor of terminating. This is indeed consistent with thread, plus it can help catch a bunch of bugs.

I think nodiscard is something we should leave for eventual LEWG discussions, since it is essentially unobservable and thus really a matter of an implementation, as I understand. Let's not spend time on debating that here.

**Comment by LeeHowes** [comment]
It is still a question whether we want the concept to require anything specific on this.

**Comment by griwes** [comment]
For the concept we *could* say it's unspecified, but I feel like specifying it to require one action or another would be helpful for generic code. Especially since it's a hard failure condition.

**Comment by brycelelbach** [comment]
We discussed this at the NVIDIA F2F meeting, and decided we actually do want to do something for cancellation in version 1.

```
// Model 1b (current status quo): <-- We prefer this over 1a
struct promise
{
   void set_value(T value);
   void set_exception(E error);
};

std::pair<promise, future> make_promise(Executor exec);
// This is one executor property.

std::pair<promise, future> make_promise(Executor exec, CancellationCallback cc); // cc is void()
// This is another executor property.

auto [p, f] = make_promise(exec);
// ...
auto g = then_execute(other_exec, f, c);
p(v); // `f` is made ready with the value `v`.
      // After this, `c(v)` is executed on `other_exec` asynchronously; `g` is made ready with the result.
```

Poll: Proceed with model 1b (model 1 with cancellation)
SF F N A SA
4 2 1 0 0

Poll: Cancellation should be in version 1 (Rapperswil) of the paper
SF F N A SA
1 4 1 1 0

Action Items:
* [ ] After we have combined the promise executor property into this paper, add the cancellation stuff as above.

**Comment by brycelelbach** [comment]
Note: We agreed we want cancellation, but we did NOT agree on what will cause cancellation - does destruction cancel? Do we need a specific cancellation mechanism?

For now, we will not include a cancellation mechanism in our future concepts.

**Comment by LeeHowes** [comment]
Closing. Text is in paper, we can tweak as general wording.
